packages/backend/server/src/plugins/copilot/providers/utils.ts

```diff
@@ -367,3 +370,109 @@ export class CitationParser {
     return footnotes.join('\n');
   }
 }
+
+export type AITools = {
+  web_search_preview: ReturnType<typeof openai.tools.webSearchPreview>;
+  web_search_exa: ReturnType<typeof createExaSearchTool>;
+  web_crawl_exa: ReturnType<typeof createExaCrawlTool>;
+};
+
+type ChunkType = TextStreamPart<AITools>['type'];
+
+export class TextStreamParser {
```

Comment:
Should TextStreamParser consider scalability?
* Chunk Type is extensible.
* Chunk Parser Algorithm is extensible.

```ts
declare 'module' {
   interface AITools {
      custom_tool: ReturnTypeOfCustomTool
   }
}
```

```ts
const parser = TextStreamParser.create({
  custom_tool: parseCustomToolChunk(chunk),
})
```